### PR TITLE
Added the ability to add custom extension functions to the VariableDictionary

### PR DESCRIPTION
**Pull request by dkarzon**
Building off the functionality used for the BuiltInFunctions this lets you add your own custom functions to the Octostache expressions without needing to fork the library yourself.

Useful for domain specific logic that falls outside of the current function selection. (See test cases for usage)

```
variables.AddExtension("supercoolfunc", ExtensionFixture.SuperCoolFunc);
var result = variables.Evaluate("#{Foo|supercoolfunc}");
```

Not entirely sure on the approach to get the Extensions dictionary from the VariableDictionary instance to the TemplateEvaluator so let me know if there is a more preferable approach.

### CODE REVIEW

**Comment by dkarzon** [comment]
@droyad @matt-richardson Any movement on this one? Would be great to be able to use the official release with these changes instead of building myself from a branch.

**Comment by dkarzon** [comment]
Bump 👋

**Comment by dkarzon** [comment]
Any chance this could get a look at?

**Comment by dkarzon** [comment]
@mjhilton Any feedback on this one?

**Comment by mjhilton** [comment]
Hey @dkarzon, thanks for your PR 🙂

Overall the idea looks good, and I like that you've handled potential conflict between BuiltInFunctions and Extensions added afterwards. Thankyou also for adding tests ✅

You might have seen I've been doing some spring cleaning in the repo related to our build and release pipelines lately; I still need to make a couple of modifications to be able to run the pipeline against your fork. I'll try to get that done in the next few days so we can get the PR green.

Once I've done that, I'll do a slightly more detailed review, but I don't anticipate any major changes.

File: source/Octostache/VariableDictionary.cs [hunk 1]
```diff
@@ -326,5 +327,16 @@ public void Add(string key, string? value)
...
         }
+
+        /// <summary>
+        /// Adds a custom extension function
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="func"></param>
+        public void AddExtension(string name, Func<string?, string[], string?> func)
+        {
+            // Naming conflicts with BuiltInFunctions?
```
Comment (mjhilton) [inline source/Octostache/VariableDictionary.cs:338]: This has been handled in `TemplateEvaluator`, so I don't think we need this comment any more

**Review by mjhilton — CHANGES_REQUESTED** [review]
Hi @dkarzon, I've run the build for your PR. The failed component was the `Enforce Code Style` stage, which failed with the following:

```
[14:51:53 ] Pending cleanups found in the following files:
[14:51:53 ]     M source/Octostache.Tests/ExtensionFixture.cs
[14:51:53 ]     M source/Octostache/Templates/EvaluationContext.cs
[14:51:53 ]     M source/Octostache/Templates/TemplateEvaluator.cs
[14:51:53 ]     M source/Octostache/VariableDictionary.cs
[14:51:53 ] Code Cleanups found
[14:51:53 ] Please run ReSharper Cleanup on your branch
```

You'll [need to run Code Cleanup locally](https://github.com/OctopusDeploy/Octostache#code-cleanup) on your branch to get it 🟢.

([Here's the `git diff`](https://github.com/OctopusDeploy/Octostache/files/8988805/gitdiff.txt) generated on the build server as a result, which is what you should end up with when you run the cleanup tool)

**Comment by dkarzon** [comment]
Thanks for the review @mjhilton I have run the code cleanup tool now. It picked up a lot of code changes unrelated to this PR which I have ignored for now and only included the files mentioned above.

**Comment by mjhilton** [comment]
The next version of Octostache as a result of this change is `3.1.0`. It has been built and published to our internal staging NuGet feed.

We have an extensive suite of integration and acceptance tests in Octopus Server that indirectly exercise Octostache. 

I've raised the PR to bump the version of Octostache in Octopus Server. Once that pipeline goes 💚, I'll approve the promotion of the new `3.1.0` to `nuget.org` and it will be available for you. Thanks for your contribution @dkarzon :)

**Comment by dkarzon** [comment]
Awesome! Thanks @mjhilton for your help 👍

**Comment by mjhilton** [comment]
🚀 [3.1.0 has been published to NuGet](https://www.nuget.org/packages/Octostache)

**Pull request merged**